### PR TITLE
Compile/run issue with Java 12 and 1.8

### LINKED ISSUE #6318: Compile/run issue with Java 12 and 1.8
FWIW: I just tried compiling Dataverse with a Java 12.02 version and while it initially looked OK, when I uploaded a file I got    
```
java.lang.NoSuchMethodError: java.nio.MappedByteBuffer.rewind()Ljava/nio/MappedByteBuffer;
        at edu.harvard.iq.dataverse.ingest.IngestableDataChecker.detectTabularDataFormat(IngestableDataChecker.java:627)
```
and related errors. This appears to be related to https://github.com/eclipse/jetty.project/issues/3244 . I tried a quick test to use the maven release attribute as discussed there but it didn't work. Possibly a newer version of the compiler plugin is needed. 

In any case, it looks like compiling with 12.02 and the current pom.xml doesn't produce a war that can be run on a Java 1.8x machine.

### PR DESCRIPTION
**Pull request by qqmyers**
## New Contributors

Welcome! New contributors should at least glance at [CONTRIBUTING.md](/CONTRIBUTING.md), especially the section on pull requests where we encourage you to reach out to other developers before you start coding. Also, please note that we measure code coverage and prefer you write unit tests. Pull requests can still be reviewed without tests or completion of the checklist outlined below. Note that we use the "closes" syntax below to trigger Github's automation to close the corresponding issue once the pull request is merged.

Thanks for your contribution to Dataverse!

## Related Issues

- closes #6318 Compile/run issue with Java 12 and 1.8

## Pull Request Checklist

- [ ] Unit [tests][] completed
- [ ] Integration [tests][]: None
- [ ] Deployment requirements, [SQL updates][], [Solr updates][], etc.: None
- [ ] [Documentation][docs] completed
- [ ] Merged latest from "develop" [branch][] and resolved conflicts

[tests]: http://guides.dataverse.org/en/latest/developers/testing.html
[SQL updates]: http://guides.dataverse.org/en/latest/developers/sql-upgrade-scripts.html
[Solr updates]: https://github.com/IQSS/dataverse/blob/develop/conf/solr/7.3.0/schema.xml
[docs]: http://guides.dataverse.org/en/latest/developers/documentation.html
[branch]: http://guides.dataverse.org/en/latest/developers/branching-strategy.html

### CODE REVIEW

**Comment by qqmyers** [comment]
I've tested this with a Java 12.02 and Java 1.8.0_112 compilers and both produce .war files that are compatible with running on a Java 8 machine (1.8.0_51).
I have not tried removing the source and target params, which are the old way of doing it. I'm not sure if those are required for a java 8 compilers or not. 
I have verified that the plugin update is required along with setting the release to 8.

**Comment by kcondon** [comment]
Did anyone try the installer? There was recently an RT ticket where someone installed with Java 12 and got a strange error. @pdurbin maybe you can do a dev sanity check?

**Comment by kcondon** [comment]
@pdurbin I could use some advice/help here: does this require java 12 to be installed on Jenkins to test this? Is this how we are planning to operate going forward? I am getting this error currently but jdk8 is still the build env:

Waiting for Jenkins to finish collecting data
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.0:compile (default-compile) on project dataverse: Fatal error compiling: invalid flag: --release -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException

I am not 100% sure how to add java 12 to Jenkins without potentially messing up other build jobs.

**Comment by qqmyers** [comment]
FWIW: I was able to compile against Java 1.8.0_112 (older than what you use). I do have maven 3.5.4 on that machine though - could it be a maven version issue?

**Comment by djbrooke** [comment]
@qqmyers @pdurbin @kcondon thanks for the discussion on this one. Sorry for the potentially ignorant question, but what do we as a project gain from this? I thought we were a bit behind with our Java versions (https://github.com/IQSS/dataverse/issues/4259), so does this represent an upgrade or something else?

**Comment by pdurbin** [comment]
@kcondon I haven't tried executing this pull request at all but my approach would be to try our "spin up" scripts. Here's where the Java version is controlled:

https://github.com/IQSS/dataverse-sample-data/blob/baf56de2a9b8b230168d6e6e1758a10201e0ed5e/ec2config.yaml#L87

@qqmyers what does `<releases>` do? I haven't looked for docs on that but if you have them handy, please post a link.

@djbrooke we never want to be stuck on any particular version of a software dependency. It's always healthy to test your software on newer versions of your dependencies to get a sense of how much effort that upgrade will be. It's an investment in the future because some day version you depend on will go end of life (EOL).

**Comment by qqmyers** [comment]
@djbrooke - I set up maven/java on a new dev machine and figured I should use the latest versions if possible. The fix here was what I needed to be able to compile /package a war that can run back on Java 1.8x. I don't know that the project needs to upgrade its infrastructure right now but I thought it would help developers at least. If it had just worked, I think updating would be useful. If there's some issue in the build chain, perhaps just having the issue and fix for others to try is enough for now.

**Comment by qqmyers** [comment]
@pdurbin - release is the new flag that replaces source and target. The later Java's don't appear to do everything needed to be 1.8 compatible with source and target (Per the URL in the issue). And, at least for me, the Java 1.8 compiler seems to ignore the release flag OK and still use source and target so both work.

**Comment by kcondon** [comment]
This is what happens on our current build env, granted we are only running java 8 so apologies if it should be obvious that java 12 is required:

Waiting for Jenkins to finish collecting data
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.0:compile (default-compile) on project dataverse: Fatal error compiling: invalid flag: --release -> [Help 1]
org.apache.maven.lifecycle.LifecycleExecutionException: Failed to execute goal org.apache.maven.plugins:maven-compiler-plugin:3.8.0:compile (default-compile) on project dataverse: Fatal error compiling
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:217)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:153)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:145)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:84)
	at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject(LifecycleModuleBuilder.java:59)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.singleThreadedBuild(LifecycleStarter.java:183)
	at org.apache.maven.lifecycle.internal.LifecycleStarter.execute(LifecycleStarter.java:161)
	at org.apache.maven.DefaultMaven.doExecute(DefaultMaven.java:320)
	at org.apache.maven.DefaultMaven.execute(DefaultMaven.java:156)
	at org.jvnet.hudson.maven3.launcher.Maven3Launcher.main(Maven3Launcher.java:117)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launchStandard(Launcher.java:329)
	at org.codehaus.plexus.classworlds.launcher.Launcher.launch(Launcher.java:239)
	at org.jvnet.hudson.maven3.agent.Maven3Main.launch(Maven3Main.java:182)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at hudson.maven.Maven3Builder.call(Maven3Builder.java:139)
	at hudson.maven.Maven3Builder.call(Maven3Builder.java:70)
	at hudson.remoting.UserRequest.perform(UserRequest.java:212)
	at hudson.remoting.UserRequest.perform(UserRequest.java:54)
	at hudson.remoting.Request$2.run(Request.java:369)
	at hudson.remoting.InterceptingExecutorService$1.call(InterceptingExecutorService.java:72)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	at java.lang.Thread.run(Thread.java:748)
Caused by: org.apache.maven.plugin.MojoExecutionException: Fatal error compiling
	at org.apache.maven.plugin.compiler.AbstractCompilerMojo.execute(AbstractCompilerMojo.java:1140)
	at org.apache.maven.plugin.compiler.CompilerMojo.execute(CompilerMojo.java:188)
	at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo(DefaultBuildPluginManager.java:101)
	at org.apache.maven.lifecycle.internal.MojoExecutor.execute(MojoExecutor.java:209)
	... 30 more
Caused by: org.codehaus.plexus.compiler.CompilerException: invalid flag: --release
	at org.codehaus.plexus.compiler.javac.JavaxToolsCompiler.compileInProcess(JavaxToolsCompiler.java:173)
	at org.codehaus.plexus.compiler.javac.JavacCompiler.performCompile(JavacCompiler.java:174)
	at org.apache.maven.plugin.compiler.AbstractCompilerMojo.execute(AbstractCompilerMojo.java:1129)
	... 33 more
Caused by: java.lang.IllegalArgumentException: invalid flag: --release
	at com.sun.tools.javac.api.JavacTool.processOptions(JavacTool.java:206)
	at com.sun.tools.javac.api.JavacTool.getTask(JavacTool.java:156)
	at com.sun.tools.javac.api.JavacTool.getTask(JavacTool.java:107)
	at com.sun.tools.javac.api.JavacTool.getTask(JavacTool.java:64)
	at org.codehaus.plexus.compiler.javac.JavaxToolsCompiler.compileInProcess(JavaxToolsCompiler.java:125)
	... 35 more
[ERROR] 
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException

**Comment by pdurbin** [comment]
> release is the new flag that replaces source and target

Oooooh, so that's what those guys were talking about when they said "-source and -target are deprecated in Java9 and forward, as far as I understand" at http://irclog.greptilian.com/friendlyjava/2019-10-09

@qqmyers thanks!

**Comment by djbrooke** [comment]
Thanks all for the discussion. I'd love to get Java upgraded, as I've previously prioritized #4259. 

Assigning to @scolapasta to determine the correct path here.

**Comment by djbrooke** [comment]
Sounds good @qqmyers. Let us know once you make those changes and I'll ask @scolapasta to take a look.

**Comment by qqmyers** [comment]
@djbrooke - using profiles does the trick. Should be ready for review.

**Comment by djbrooke** [comment]
Thanks @pdurbin @sekmiller @scolapasta for the quick explanation. I understand now that this is not a Java 12 upgrade. The title of the PR was throwing me. 

@kcondon, so if I understand correctly it becomes a question of whether or not we need to upgrade our local Jenkins to a new version of Maven to support this? I think since we're eventually going to upgrade our Java/JavaEE version it makes sense to do so. Happy to discuss in more detail.

**Comment by qqmyers** [comment]
@djbrooke @kcondon - I don't think a maven update is needed - at least the initial error reported was from my bad code (my test with a Java 8 compiler was flawed, once I fixed it, I saw the same error). The updated code in the PR should work for both 1.8 and later compilers now and, unless there's a new error, shouldn't require any maven/other updates to the IQSS build process.

File: pom.xml [hunk 1]
```diff
@@ -798,5 +787,46 @@
...
+                <compilerArgument>${compilerArgument}</compilerArgument>
+              </configuration>
+            </plugin>
+          </plugins>
+        </build>
+      </profile>
+      <profile>
+        <id>Java9Plus</id>
+        <activation>
+          <jdk>[1.9</jdk>
```
Comment (pdurbin) [inline pom.xml]: Is this `[` supposed to be there? It says `<jdk>[1.9</jdk>`.

**Pull request merged**